In [ ]:
!python -m pip install -U python-woc pandas matplotlib
# please clear the output of this cell in your notebook before checking it in

In [5]:
from woc.remote import WocMapsRemote
from tqdm import tqdm
import pandas as pd
# creates the client
woc = WocMapsRemote( base_url="https://worldofcode.org/api/")
# if you got an API key
# woc = WocMapsRemote( base_url="https://worldofcode.org/api/", api_key="woc-XXXXXX-YYYYYY" )

# Now for each of the ten projects you were assigned get commits, e.g.
projects = ['epam/med3web', 'indralab/indra_db', 'choosehappy/HistoQC', 'qulacs/qulacs', 'ropensci/stplanr', 'ZhaoJ9014/face.evoLVe',
            'rasbt/mlxtend', 'systerel/S2OPC', 'teusH/MySense', 'kuwisdelu/Cardinal']
list_df_commits = []
for prj in projects:
  # Convert GitHub repository names to WoC V2412 project names
  prj = prj.lower().replace('/','_',2)
  commits = woc.get_values('p2c', prj)
  df = pd.DataFrame(commits, columns=["sha1"])
  df['project'] = prj
  list_df_commits.append(df)
df = pd.concat(list_df_commits)
#perhaps save the list (if no errors) so you do not need to retrieve them again
df.to_csv('df_commits.csv')
df.head(1)

,sha1,project
0,00225b0980e17ccd8abb39571feecbbb98f9e60f,epam_med3web


In [6]:
# If the number of commits its not very large, you can try to get them all at the same time,
# The max batch size is 10
import time
# let us first split df['sha1'] in chunks
chunks = [df['sha1'][x:x+10] for x in range(0, len(df), 10)]
commit_data = []

for chunk in tqdm(chunks): # iterate over the commits
  # res, err = woc.show_content_many('commit',chunk.to_list())

  # commit.tch returns the same data but faster
  res, err = woc.get_values_many('commit.tch',chunk.to_list())
  res = {k: v[0] for k, v in res.items()}  # this conversion is necessary because of the internal implementation

  # to walk around the rate limit
  time.sleep(1)

  if err: # check for errors
    print('Got Errors', err)

  for commit_sha, commit in res.items():
    # flatten commit objects
    commit_data.append({
          'commit': commit_sha,
          'tree': commit[0],
          'parent': list(commit[1]),
          'author': commit[2][0],
          'author_time': int(commit[2][1]),
          'author_tz': commit[2][2],
          'committer': commit[3][0],
          'committer_time': int(commit[3][1]),
          'committer_tz': commit[3][2],
          'message': commit[4],
    })

df_commit_data = pd.DataFrame(commit_data)
df_commit_data = df_commit_data.merge(df, left_on='commit', right_on='sha1')
df_commit_data.to_csv('df_commit_data.csv', index=False)
df_commit_data.head(2)

  4%|███▎                                                                           | 137/3230 [02:19<52:30,  1.02s/it]

Got Errors {'027e8d769e0bcc38842ee526d2fa3fa55d0dbdc6': 'Key 027e8d769e0bcc38842ee526d2fa3fa55d0dbdc6 not found in /da5_fast/All.sha1c/commit_2.tch'}


  8%|██████▎                                                                        | 257/3230 [04:21<50:27,  1.02s/it]

Got Errors {'526338b1a28eeeca5bfed5695735e4d222d4077d': 'Key 526338b1a28eeeca5bfed5695735e4d222d4077d not found in /da5_fast/All.sha1c/commit_82.tch'}


 13%|██████████▏                                                                    | 418/3230 [07:05<47:45,  1.02s/it]

Got Errors {'bb0a72b5c744fffb187e3b7db913e383f9709bbb': 'Key bb0a72b5c744fffb187e3b7db913e383f9709bbb not found in /da5_fast/All.sha1c/commit_59.tch'}


 20%|███████████████▉                                                               | 651/3230 [11:03<43:55,  1.02s/it]

Got Errors {'2f2a2fa5e8a58e8a40ec7a7c842b03c56c7e9438': 'Key 2f2a2fa5e8a58e8a40ec7a7c842b03c56c7e9438 not found in /da5_fast/All.sha1c/commit_47.tch'}


 20%|████████████████                                                               | 655/3230 [11:07<43:45,  1.02s/it]

Got Errors {'320cf426fd7f7f90100c40c193979ef02eee302e': 'Key 320cf426fd7f7f90100c40c193979ef02eee302e not found in /da5_fast/All.sha1c/commit_50.tch'}


 20%|████████████████                                                               | 658/3230 [11:10<43:42,  1.02s/it]

Got Errors {'33212f76141c1967fe941f9d7cd1ac361ea0dcee': 'Key 33212f76141c1967fe941f9d7cd1ac361ea0dcee not found in /da5_fast/All.sha1c/commit_51.tch'}


 21%|████████████████▏                                                              | 664/3230 [11:16<43:34,  1.02s/it]

Got Errors {'38696575511ccda0fa3158e71c09017613c564f6': 'Key 38696575511ccda0fa3158e71c09017613c564f6 not found in /da5_fast/All.sha1c/commit_56.tch'}


 22%|█████████████████                                                              | 698/3230 [11:51<42:57,  1.02s/it]

Got Errors {'50dd230e13be4dcef91078de3978994d5ff1e155': 'Key 50dd230e13be4dcef91078de3978994d5ff1e155 not found in /da5_fast/All.sha1c/commit_80.tch'}


 23%|█████████████████▊                                                             | 730/3230 [12:23<42:27,  1.02s/it]

Got Errors {'662ad9b728a38ac41d5ed78f673dc3353593236b': 'Key 662ad9b728a38ac41d5ed78f673dc3353593236b not found in /da5_fast/All.sha1c/commit_102.tch'}


 25%|███████████████████▌                                                           | 800/3230 [13:35<41:15,  1.02s/it]

Got Errors {'9335bb12e6c95f9e2033a33cfc2ec45fcb9e9ca0': 'Key 9335bb12e6c95f9e2033a33cfc2ec45fcb9e9ca0 not found in /da5_fast/All.sha1c/commit_19.tch'}


 26%|████████████████████▋                                                          | 844/3230 [14:19<40:31,  1.02s/it]

Got Errors {'aeda9c2ae738ff2f1a83094ef895c222398f93d9': 'Key aeda9c2ae738ff2f1a83094ef895c222398f93d9 not found in /da5_fast/All.sha1c/commit_46.tch'}


 29%|███████████████████████▏                                                       | 948/3230 [16:05<38:45,  1.02s/it]

Got Errors {'f5b9f52be4bda160751e2dde96aee3bf1bca0218': 'Key f5b9f52be4bda160751e2dde96aee3bf1bca0218 not found in /da5_fast/All.sha1c/commit_117.tch'}


 34%|██████████████████████████▍                                                   | 1096/3230 [18:36<36:13,  1.02s/it]

Got Errors {'7bf2b88d3a25a03f44f0932b1fca324b67d886a4': 'Key 7bf2b88d3a25a03f44f0932b1fca324b67d886a4 not found in /da5_fast/All.sha1c/commit_123.tch'}


 41%|███████████████████████████████▊                                              | 1318/3230 [22:22<32:26,  1.02s/it]

Got Errors {'02884731000cbc3de4f1065a8d32526aaf6d9f77': 'Key 02884731000cbc3de4f1065a8d32526aaf6d9f77 not found in /da5_fast/All.sha1c/commit_2.tch'}


 41%|███████████████████████████████▉                                              | 1323/3230 [22:27<32:18,  1.02s/it]

Got Errors {'06e34f388437f08658c68bcf32677505ce7932c5': 'Key 06e34f388437f08658c68bcf32677505ce7932c5 not found in /da5_fast/All.sha1c/commit_6.tch'}


 47%|████████████████████████████████████▋                                         | 1521/3230 [25:49<28:59,  1.02s/it]

Got Errors {'a0af44fa5aeb44c39f85c455ca941804323d366c': 'Key a0af44fa5aeb44c39f85c455ca941804323d366c not found in /da5_fast/All.sha1c/commit_32.tch'}


 47%|████████████████████████████████████▊                                         | 1523/3230 [25:51<29:06,  1.02s/it]

Got Errors {'a19d13c5ed44accef772e43a16720110069a5be6': 'Key a19d13c5ed44accef772e43a16720110069a5be6 not found in /da5_fast/All.sha1c/commit_33.tch'}


 47%|█████████████████████████████████████                                         | 1534/3230 [26:02<28:47,  1.02s/it]

Got Errors {'aab4e3ad7b7ac5cfaa3e5fbe202b9f500e0f88e3': 'Key aab4e3ad7b7ac5cfaa3e5fbe202b9f500e0f88e3 not found in /da5_fast/All.sha1c/commit_42.tch'}


 52%|████████████████████████████████████████▍                                     | 1674/3230 [28:25<26:25,  1.02s/it]

Got Errors {'055c642d8f13b1e92f9733f5d49ff571e65c02ab': 'Key 055c642d8f13b1e92f9733f5d49ff571e65c02ab not found in /da5_fast/All.sha1c/commit_5.tch'}


 57%|████████████████████████████████████████████▏                                 | 1831/3230 [31:05<23:43,  1.02s/it]

Got Errors {'228b870143211b65520b65545bb03ee5a7485cfe': 'Key 228b870143211b65520b65545bb03ee5a7485cfe not found in /da5_fast/All.sha1c/commit_34.tch'}


 65%|██████████████████████████████████████████████████▊                           | 2106/3230 [35:45<19:06,  1.02s/it]

Got Errors {'57aabbc4924154a096137d56cbfbb0690e6c7160': 'Key 57aabbc4924154a096137d56cbfbb0690e6c7160 not found in /da5_fast/All.sha1c/commit_87.tch'}


 69%|██████████████████████████████████████████████████████                        | 2239/3230 [38:00<16:50,  1.02s/it]

Got Errors {'70d47e78c54eaed5da5bee718bb18325382897b6': 'Key 70d47e78c54eaed5da5bee718bb18325382897b6 not found in /da5_fast/All.sha1c/commit_112.tch'}


 71%|███████████████████████████████████████████████████████                       | 2281/3230 [38:43<16:09,  1.02s/it]

Got Errors {'788494ce9bf1071315680fcc84ea5c74e1f04c2a': 'Key 788494ce9bf1071315680fcc84ea5c74e1f04c2a not found in /da5_fast/All.sha1c/commit_120.tch'}


 74%|█████████████████████████████████████████████████████████▋                    | 2387/3230 [40:31<14:18,  1.02s/it]

Got Errors {'8ce29a5da55e95fbd1ce2c97a3e2d60b0438f16b': 'Key 8ce29a5da55e95fbd1ce2c97a3e2d60b0438f16b not found in /da5_fast/All.sha1c/commit_12.tch'}


 76%|███████████████████████████████████████████████████████████                   | 2444/3230 [41:29<13:20,  1.02s/it]

Got Errors {'97d4cbda392fcca2962d30dc4f110e38022b35d6': 'Key 97d4cbda392fcca2962d30dc4f110e38022b35d6 not found in /da5_fast/All.sha1c/commit_23.tch'}


 77%|███████████████████████████████████████████████████████████▋                  | 2474/3230 [42:00<12:50,  1.02s/it]

Got Errors {'9d991b4afb539f3f42d0d11f486abb11c5e9a5b7': 'Key 9d991b4afb539f3f42d0d11f486abb11c5e9a5b7 not found in /da5_fast/All.sha1c/commit_29.tch'}


 78%|████████████████████████████████████████████████████████████▊                 | 2516/3230 [42:43<12:07,  1.02s/it]

Got Errors {'a548b68f93a70679478e599ded98e38f5992ff46': 'Key a548b68f93a70679478e599ded98e38f5992ff46 not found in /da5_fast/All.sha1c/commit_37.tch'}


 83%|████████████████████████████████████████████████████████████████▋             | 2677/3230 [45:27<09:22,  1.02s/it]

Got Errors {'c3a8183485f737fb1a20cb86f6f74ada5b9d5422': 'Key c3a8183485f737fb1a20cb86f6f74ada5b9d5422 not found in /da5_fast/All.sha1c/commit_67.tch'}


 84%|█████████████████████████████████████████████████████████████████▋            | 2722/3230 [46:12<08:37,  1.02s/it]

Got Errors {'cc4ee650e5535e88dc911671858aacb3137327dc': 'Key cc4ee650e5535e88dc911671858aacb3137327dc not found in /da5_fast/All.sha1c/commit_76.tch'}


 86%|███████████████████████████████████████████████████████████████████           | 2778/3230 [47:09<07:40,  1.02s/it]

Got Errors {'d73460b2d212924b54bc0d2c187cd73be0e20671': 'Key d73460b2d212924b54bc0d2c187cd73be0e20671 not found in /da5_fast/All.sha1c/commit_87.tch'}


 87%|███████████████████████████████████████████████████████████████████▌          | 2799/3230 [47:31<07:19,  1.02s/it]

Got Errors {'db770a51cd48ce17e3666a2a7a28a91c7ef10ca6': 'Key db770a51cd48ce17e3666a2a7a28a91c7ef10ca6 not found in /da5_fast/All.sha1c/commit_91.tch'}


 88%|████████████████████████████████████████████████████████████████████▉         | 2857/3230 [48:30<06:19,  1.02s/it]

Got Errors {'e67eadec3a240169a04f957ea480dba1fedc87c7': 'Key e67eadec3a240169a04f957ea480dba1fedc87c7 not found in /da5_fast/All.sha1c/commit_102.tch'}


 89%|█████████████████████████████████████████████████████████████████████▏        | 2864/3230 [48:37<06:14,  1.02s/it]

Got Errors {'e7df6180995f4282e8faccf80da76ccc57b852c6': 'Key e7df6180995f4282e8faccf80da76ccc57b852c6 not found in /da5_fast/All.sha1c/commit_103.tch'}


 91%|██████████████████████████████████████████████████████████████████████▊       | 2933/3230 [49:47<05:02,  1.02s/it]

Got Errors {'f4e473f900def718a636a89c4fd047b6ecfa912a': 'Key f4e473f900def718a636a89c4fd047b6ecfa912a not found in /da5_fast/All.sha1c/commit_116.tch'}


 93%|████████████████████████████████████████████████████████████████████████▏     | 2989/3230 [50:44<04:05,  1.02s/it]

Got Errors {'fffe0301e3a9b20ffe2ef00f1dc349366a37409a': 'Key fffe0301e3a9b20ffe2ef00f1dc349366a37409a not found in /da5_fast/All.sha1c/commit_127.tch'}


 96%|███████████████████████████████████████████████████████████████████████████▏  | 3114/3230 [52:52<01:58,  1.02s/it]

Got Errors {'e6d557ff705c2fdcb815921fee93c7ec4e83fba1': 'Key e6d557ff705c2fdcb815921fee93c7ec4e83fba1 not found in /da5_fast/All.sha1c/commit_102.tch'}


100%|██████████████████████████████████████████████████████████████████████████████| 3230/3230 [54:50<00:00,  1.02s/it]


,commit,tree,parent,author,author_time,author_tz,committer,committer_time,committer_tz,message,sha1,project
0,00225b0980e17ccd8abb39571feecbbb98f9e60f,88f5ec8f327d5ca7a6ab1daf6901be26ae7574bd,[bb72c3617d90c63082596d5e65356398f26a3e81],Vladislav Shubnikov <Vladislav_Shubnikov@epam....,1526638320,+0300,Vladislav Shubnikov <Vladislav_Shubnikov@epam....,1526638320,+0300,add ActiveVolume functionality (skull detect a...,00225b0980e17ccd8abb39571feecbbb98f9e60f,epam_med3web
1,0042d7f2465621277a4d11e5995304390e1c82b6,a21c656d6f845face56943d969638723a8ef9281,"[313f0a1e535973004941a8ae0664fb19e95d0cc3, 6c2...",DanilRostov <31970322+DanilRostov@users.norepl...,1684534336,-0700,GitHub <noreply@github.com>,1684534336,-0700,Merge pull request #153 from epam/feature/impr...,0042d7f2465621277a4d11e5995304390e1c82b6,epam_med3web


In [7]:
# what does the first record looks like?
# commit  parents
v = commit_data[0]
# commit tree, [parent{s}], [author, unixtime, tz ], [ committer, unixtime, tz ], Commit_message ]
print(v)

{'commit': '00225b0980e17ccd8abb39571feecbbb98f9e60f', 'tree': '88f5ec8f327d5ca7a6ab1daf6901be26ae7574bd', 'parent': ['bb72c3617d90c63082596d5e65356398f26a3e81'], 'author': 'Vladislav Shubnikov <Vladislav_Shubnikov@epam.com>', 'author_time': 1526638320, 'author_tz': '+0300', 'committer': 'Vladislav Shubnikov <Vladislav_Shubnikov@epam.com>', 'committer_time': 1526638320, 'committer_tz': '+0300', 'message': 'add ActiveVolume functionality (skull detect and remove) in early development stage (will be migrated from c source)\n\n- Add basic geometry class for active surface iterations\n- Fix some coding standard issues\n'}


In [8]:
# Now that the data is retrieved, save it and commit to your GH fork
# First, we need to flatten info in order to export as csv
# our dataframe will have columns project,'commit, author, time, message'
dfinf = pd.DataFrame(columns=['project', 'commit', 'author', 'time', 'message'])
for k in commit_data:
  project = df.loc[df['sha1'] == k['commit'], 'project'].iloc[0]
  row = pd.Series({'project':project, 'commit': k['commit'], 'author': k['author'], 'time': k['author_time'], 'message':k['message']})
  dfinf = pd.concat([dfinf, row.to_frame().T ], ignore_index=True)


In [12]:
# check if it has the right content
dfinf.head(1)

,project,commit,author,time,message
0,epam_med3web,00225b0980e17ccd8abb39571feecbbb98f9e60f,Vladislav Shubnikov <Vladislav_Shubnikov@epam....,1526638320,add ActiveVolume functionality (skull detect a...


In [16]:
#mode a means append, so you have all your projects in the same file
yournetid='pvickery'
dfinf.to_csv(yournetid+'_project_summary.csv', index=False,sep=';', mode='a', header=False)

In [5]:

project_summary = df_commit_data.groupby('project').agg(
    commits=('commit', 'count'),
    authors=('author', 'nunique'),
    max_time=('author_time', 'max'),
    min_time=('author_time', 'min')
).reset_index()
project_summary['max_time'] = pd.to_datetime(
    project_summary['max_time'], unit='s', utc=True
)

project_summary['min_time'] = pd.to_datetime(
    project_summary['min_time'], unit='s', utc=True
)
project_summary

,project,commits,authors,max_time,min_time
0,choosehappy_histoqc,610,43,2026-04-13 20:00:04+00:00,2017-10-16 13:29:10+00:00
1,epam_med3web,1327,79,2025-08-21 01:14:36+00:00,2017-12-25 15:02:37+00:00
2,indralab_indra_db,3888,24,2025-10-16 21:51:36+00:00,2018-09-20 21:16:31+00:00
3,kuwisdelu_cardinal,1036,29,2025-10-14 18:19:22+00:00,2013-11-17 11:46:32+00:00
4,qulacs_qulacs,3803,113,2025-09-01 00:44:13+00:00,2018-10-05 05:24:32+00:00
5,rasbt_mlxtend,3288,213,2026-02-27 08:53:23+00:00,2014-08-14 01:56:16+00:00
6,ropensci_stplanr,2753,50,2025-05-03 08:48:58+00:00,2015-01-30 08:36:50+00:00
7,systerel_s2opc,13436,51,2025-11-03 13:31:44+00:00,2016-03-14 15:11:00+00:00
8,teush_mysense,1364,6,2025-04-09 15:25:56+00:00,2017-02-22 14:51:50+00:00
9,zhaoj9014_face.evolve,750,11,2025-03-20 02:45:06+00:00,2019-01-04 02:22:17+00:00


# Make sure you check in to your fork not just the notebook but also the csv files!!!

# Don't forget to add requested data from github and this notebook
### For each of the 10 projects go to their github repo and get the number of stars, number of forks, and the last commit date
### Report (in your notebook) the number of commits, the number of authors, and max and min time for each project based on WoC commits and also add the info you obtained from github

| Project | # Commits (WoC) | # Authors (WoC) | Max Time (WoC) | Min Time (WoC) | GitHub Stars | GitHub Forks | GitHub Last Commit |
|---|---:|---:|---|---|---:|---:|---|
| choosehappy_histoqc | 610 | 43 | 2026-04-13 20:00:04+00:00 | 2017-10-16 13:29:10+00:00 | 331 | 118 | September 16, 2026 |
| epam_med3web | 1327 | 79 | 2025-08-21 01:14:36+00:00 | 2017-12-25 15:02:37+00:00 | 368 | 105 | September 24, 2026 |
| indralab_indra_db | 3888 | 24 | 2025-10-16 21:51:36+00:00 | 2018-09-20 21:16:31+00:00 | 22 | 12 | July 9, 2026 |
| kuwisdelu_cardinal | 1036 | 29 | 2025-10-14 18:19:22+00:00 | 2013-11-17 11:46:32+00:00 | 81 | 18 | July 30, 2026 |
| qulacs_qulacs | 3803 | 113 | 2025-09-01 00:44:13+00:00 | 2018-10-05 05:24:32+00:00 | 529 | 147 | September 29, 2026 |
| rasbt_mlxtend | 3288 | 213 | 2026-02-27 08:53:23+00:00 | 2014-08-14 01:56:16+00:00 | 5200 | 920 | September 29, 2026 |
| ropensci_stplanr | 2753 | 50 | 2025-05-03 08:48:58+00:00 | 2015-01-30 08:36:50+00:00 | 443 | 68 | April 17, 2025 |
| systerel_s2opc | 13436 | 51 | 2025-11-03 13:31:44+00:00 | 2016-03-14 15:11:00+00:00 | 28 | 6 | September 30, 2026 |
| teush_mysense | 1364 | 6 | 2025-04-09 15:25:56+00:00 | 2017-02-22 14:51:50+00:00 | 71 | 18 | September 1, 2026 |
| zhaoj9014_face.evolve | 750 | 11 | 2025-03-20 02:45:06+00:00 | 2019-01-04 02:22:17+00:00 | 3600 | 759 | March 19, 2025 |
